### PR TITLE
Fixed C4QueryObserver (LiteCore's LiveQuery)

### PR DESCRIPTION
**Pull request by snej**
I reported the issue as "C4QueryObserver does not trigger when a doc is purged", but that turned out not to be the case; the problem is a race condition that makes it miss some changes. To fix that I changed the way that it gets triggered: now it's only on transaction commits, not on individual changes.

**NOTE:** It's probably best to read through the individual commits rather than the merged view.

Fixes CBL-688

### CODE REVIEW

**Comment by borrrden** [comment]
I'll remove myself for now but when the tests are settled then add me back.

File: C/c4PredictiveQuery.cc [hunk 1]
```diff
@@ -59,15 +60,23 @@ class C4PredictiveModelInternal : public PredictiveModel {
...
 
+#endif // COUCHBASE_ENTERPRISE
+
 
 void c4pred_registerModel(const char *name, C4PredictiveModel model) C4API {
+#ifdef COUCHBASE_ENTERPRISE
     auto context = retained(new C4PredictiveModelInternal(model));
     context->registerAs(name);
+#else
+    abort();
```
Comment (jayahariv) [inline C/c4PredictiveQuery.cc]: when fails, should we need to include an error message for easy debugging?

File: C/c4PredictiveQuery.cc [hunk 2]
```diff
@@ -59,15 +60,23 @@ class C4PredictiveModelInternal : public PredictiveModel {
...
+#endif
 }
 
 
 bool c4pred_unregisterModel(const char *name) C4API {
+#ifdef COUCHBASE_ENTERPRISE
     return PredictiveModel::unregister(name);
+#else
+    abort();
```
Comment (jayahariv) [inline C/c4PredictiveQuery.cc]: same, error log?

File: C/tests/c4Test.cc [hunk 3]
```diff
@@ -159,6 +159,18 @@ void CheckError(C4Error error,
...
+        if (predicate())
+            return;
+        this_thread::sleep_for(chrono::milliseconds(100));
+    }
+    FAIL("Wait timed out after " << timeoutMillis << "ms");
+}
+
+
+
+
```
Comment (jayahariv) [inline C/tests/c4Test.cc]: extra lines?

File: LiteCore/Database/LiveQuerier.cc [hunk 4]
```diff
@@ -41,24 +51,83 @@ namespace litecore {
     ,_continuous(continuous)
     ,_delegate(delegate)
     {
-        logInfo("Created on Query %p", query);
+        logInfo("Created on Query %s", query->loggingName().c_str());
```
Comment (jayahariv) [inline LiteCore/Database/LiveQuerier.cc]: is this `loggingName` or `loggingIdentifier`?

File: C/c4PredictiveQuery.cc (hunk 1 above)
Comment (snej) [reply C/c4PredictiveQuery.cc]: ok.

File: LiteCore/Database/LiveQuerier.cc (hunk 4 above)
Comment (snej) [reply LiteCore/Database/LiveQuerier.cc]: `loggingName` is correct. It returns the `{Query#123}` string that shows up in logs.

**Comment by borrrden** [comment]
There is something wrong with this.  I think the CE version is missing an include (maybe it is guarded by `COUCHBASE_ENTERPRISE` or something)

**Pull request merged**